### PR TITLE
IGVF-701-control-measurement-set

### PR DESCRIPTION
**Pull request by khzlin**

### CODE REVIEW

File: src/igvfd/schemas/measurement_set.json [hunk 1]
```diff
@@ -120,6 +120,20 @@
             "type": "string",
             "format": "uri",
             "pattern": "^https?://github\\.com/IGVF/seqspec/blob/main/assays/\\S+/spec\\.yaml$"
+        },
+        "control_file_sets": {
+            "title": "Controls",
+            "description": "A fileset that contain files that can serve as scientific controls for this measurement_set.",
```
Comment (gabdank) [inline src/igvfd/schemas/measurement_set.json:126]: File sets that can serve as ....

File: src/igvfd/schemas/measurement_set.json [hunk 2]
```diff
@@ -120,6 +120,20 @@
...
+            "type": "array",
+            "default": [],
+            "uniqueItems": true,
+            "items": {
+                "title": "Control",
+                "description": "A fileset that contains files that can serve as a scientific control for this measurement_set.",
```
Comment (gabdank) [inline src/igvfd/schemas/measurement_set.json:132]: A file set ....

File: src/igvfd/schemas/measurement_set.json [hunk 3]
```diff
@@ -120,6 +120,20 @@
             "type": "string",
             "format": "uri",
             "pattern": "^https?://github\\.com/IGVF/seqspec/blob/main/assays/\\S+/spec\\.yaml$"
+        },
+        "control_file_sets": {
+            "title": "Controls",
+            "description": "A fileset that contain files that can serve as scientific controls for this measurement_set.",
+            "type": "array",
```
Comment (gabdank) [inline src/igvfd/schemas/measurement_set.json:127]: need to be minItem = 1

File: src/igvfd/types/file_set.py [hunk 4]
```diff
@@ -41,6 +42,18 @@ class FileSet(Item):
...
+        'title': 'MeasurementSets this FileSets act as a control.',
+        'type': 'array',
+        'items': {
+            'type': ['string', 'object'],
+            'linkFrom': 'FileSet.control_file_sets',
+        },
+        'notSubmittable': True
+    })
+    def controlling_file_sets(self, request, controlling_file_sets):
+        return paths_filtered_by_status(request, controlling_file_sets)
```
Comment (gabdank) [inline src/igvfd/types/file_set.py:55]: title: File sets controlled by this file set. 
rev-link name: "controlled_file_sets"

File: src/igvfd/types/file_set.py [hunk 5]
```diff
@@ -26,7 +26,8 @@ class FileSet(Item):
     name_key = 'accession'
     schema = load_schema('igvfd:schemas/file_set.json')
     rev = {
-        'files': ('File', 'file_set')
+        'files': ('File', 'file_set'),
+        'controlling_file_sets': ('FileSet', 'control_file_sets')
```
Comment (gabdank) [inline src/igvfd/types/file_set.py:30]: controlled_file_sets

File: src/igvfd/schemas/changelogs/measurement_set.md [hunk 6]
```diff
@@ -3,6 +3,8 @@
 ### Minor changes since schema version 4
 
 * Add `files`.
+* Add `control_file_sets`.
+* Add `controlling_file_sets`.
```
Comment (gabdank) [inline src/igvfd/schemas/changelogs/measurement_set.md:7]: controlled_file_sets

File: src/igvfd/schemas/changelogs/curated_set.md [hunk 7]
```diff
@@ -5,6 +5,7 @@
 * Add `files`.
 * Add `dbxrefs`.
 * Add `guide RNAs` to `curated_set_type`.
+* Add `controlling_file_sets`.
```
Comment (gabdank) [inline src/igvfd/schemas/changelogs/curated_set.md:8]: controlled_file_sets

File: src/igvfd/mappings/measurement_set.json [hunk 8]
```diff
@@ -314,6 +314,12 @@
                     "construct_libraries": {
                         "type": "keyword"
                     },
+                    "control_file_sets": {
+                        "type": "keyword"
+                    },
+                    "controlling_file_sets": {
```
Comment (gabdank) [inline src/igvfd/mappings/measurement_set.json:320]: controlled_file_sets

**Review by gabdank — CHANGES_REQUESTED** [review]
need also update the facets and columns in search/configs

File: src/igvfd/mappings/measurement_set.json (hunk 8 above)
Comment (khzlin) [reply src/igvfd/mappings/measurement_set.json:320]: fixed

File: src/igvfd/schemas/changelogs/curated_set.md (hunk 7 above)
Comment (khzlin) [reply src/igvfd/schemas/changelogs/curated_set.md:8]: fixed

File: src/igvfd/schemas/changelogs/measurement_set.md (hunk 6 above)
Comment (khzlin) [reply src/igvfd/schemas/changelogs/measurement_set.md:7]: fixed

File: src/igvfd/schemas/measurement_set.json (hunk 1 above)
Comment (khzlin) [reply src/igvfd/schemas/measurement_set.json:126]: fixed

File: src/igvfd/schemas/measurement_set.json (hunk 2 above)
Comment (khzlin) [reply src/igvfd/schemas/measurement_set.json:132]: fixed

File: src/igvfd/schemas/measurement_set.json (hunk 3 above)
Comment (khzlin) [reply src/igvfd/schemas/measurement_set.json:127]: fixed

File: src/igvfd/types/file_set.py (hunk 5 above)
Comment (khzlin) [reply src/igvfd/types/file_set.py:30]: fixed

File: src/igvfd/types/file_set.py (hunk 4 above)
Comment (khzlin) [reply src/igvfd/types/file_set.py:55]: rev-link name: control_of
fixed

**Review by gabdank — CHANGES_REQUESTED** [review]
please update inserts with the new relationship

File: src/igvfd/schemas/measurement_set.json [hunk 9]
```diff
@@ -120,6 +120,21 @@
             "type": "string",
             "format": "uri",
             "pattern": "^https?://github\\.com/IGVF/seqspec/blob/main/assays/\\S+/spec\\.yaml$"
+        },
+        "control_file_sets": {
+            "title": "Controls",
+            "description": "File sets that can serve as scientific controls for this measurement_set.",
+            "type": "array",
+            "minimum": 1,
+            "default": [],
```
Comment (gabdank) [inline src/igvfd/schemas/measurement_set.json:129]: can not be default empty if minimum is 1 item

**Review by gabdank — CHANGES_REQUESTED** [review]
default can not be empty if we require one item inside

**Review by gabdank — DISMISSED** [review]

File: src/igvfd/schemas/measurement_set.json (hunk 9 above)
Comment (khzlin) [reply src/igvfd/schemas/measurement_set.json]: done

**Pull request merged**